w3af/core/data/parsers/doc/wsdl.py

```diff
@@ -19,194 +19,110 @@
-        return self._ns
-
-    def set_ns(self, namespace):
-        self._ns = namespace
-
-    def get_type(self):
-        return self._type
-
-    def set_type(self, paramType):
-        self._type = paramType
-
-
-class remoteMethod:
-    """
-    This class represents a remote method call.
-    """
-    def __init__(self):
-        self._name = ''
-        self._action = ''
-        self._namespace = ''
-        self._inParameters = None
-        self._location = ''
-
-    def get_methodName(self):
-        return self._name
-
-    def set_methodName(self, name):
-        self._name = name
-
-    def get_action(self):
-        return self._action
-
-    def set_action(self, action):
-        self._action = action
-
-    def get_location(self):
-        return self._location
-
-    def set_location(self, location):
-        self._location = location
-
-    def get_namespace(self):
-        return self._namespace
-
-    def set_namespace(self, namespace):
-        self._namespace = namespace
-
-    def get_parameters(self):
-        return self._inParameters
-
-    def set_parameters(self, inparams):
-        self._inParameters = inparams
+        if not wsdl_client.wsdl.services:
+            exception_description = (
+                "The result of url: {} seems not to be valid WSDL".format(
+                    url,
+                )
+            )
+            output_manager.out.debug(exception_description)
+            return False
+        return True
+
+    def parse(self):
+        for service_name in self._wsdl_client.wsdl.services:
+            ports = self._wsdl_client.wsdl.services[service_name].ports
+            for _, port in ports.items():
+                operation_url = port.binding_options['address']
+                self._discovered_urls.add(URL(operation_url))
+
+    get_references_of_tag = BaseParser._return_empty_list
+    get_forms = BaseParser._return_empty_list
+    get_comments = BaseParser._return_empty_list
+    get_meta_redir = BaseParser._return_empty_list
+    get_meta_tags = BaseParser._return_empty_list
+    get_emails = BaseParser._return_empty_list
+
+    def get_references(self):
+        self._report_wsdl_dump()
+        return list(self._discovered_urls), []
+
+    @contextlib.contextmanager
+    def _redirect_stdout(self, new_stdout):
+        old_stdout = sys.stdout
+        try:
+            sys.stdout = new_stdout
+            yield
+        finally:
+            sys.stdout = old_stdout
+
+    def _report_wsdl_dump(self):
+        dump_capturer = StringIO()
+        with self._redirect_stdout(dump_capturer):
+            self._wsdl_client.wsdl.dump()
+        dump_string = dump_capturer.getvalue()
+        dump_info = Info(
+            name='SOAP details',
+            desc=dump_string,
+            response_ids=[],
+            plugin_name='web_spider',
+        )
+        kb.kb.append(self, 'soap_actions', dump_info)
+
+    @staticmethod
+    def get_name():
+        return 'wsdl_parser'
```

Comment:
Maybe I read too quick, but where is the parser returning the requests? You should have something similar to:

https://github.com/andresriancho/w3af/blob/master/w3af/core/data/parsers/doc/open_api/main.py#L201
https://github.com/andresriancho/w3af/blob/master/w3af/core/data/parsers/doc/open_api/main.py#L250

Reply:
While `OpenAPI().get_api_calls()` is a special method used only by `open_api` crawl plugin I've resigned from SOAP plugin according to [this comment](https://github.com/andresriancho/w3af/pull/18645#discussion_r454315477). So `wsdl` parser only provides the standard `get_references` method which is inherited from `BaseParser` and used by `web_spider`.  `get_references` method provides results as `URL` instances not `FuzzableRequest` instances. Anyway I guess that in my current flow `WSDLMutant` is missing.

Please let me know if above is the right approach.

Reply:
If your parser only returns URLs, then unless there is a SOAP endpoint which takes query string parameters... you won't be able to find any vulnerabilities based on the output of this parser.

I strongly recommend you add some code that will allow you to retrieve fuzzable requests (or was it mutants? I forgot) from the parser.

Reply:
Maybe you might need a custom plugin after all? OpenAPI document parser has its own plugin to extract the HTTP requests

https://github.com/andresriancho/w3af/blob/feature/js/w3af/plugins/crawl/open_api.py
https://github.com/andresriancho/w3af/blob/feature/js/w3af/plugins/crawl/open_api.py#L217

Now that we might have two parsers (openapi, wsdl) with their own plugins (eeek), maybe it is time to do an analysis and understand if we need to generalize all this into `web_spider`? Not sure... but worth thinking about it

Reply:
Fixed in https://github.com/andresriancho/w3af/pull/18645/commits/1a063e2341f2efc439f5f38107ba35ad577e2a28. I've added method `get_fuzzable_requests` to `BaseParser` as `WSDLParser` is second parser (after `OpenAPI`) which will produce fuzzable requests. `BaseParser.get_fuzzable_requests` returns empty list by default instead of raising `NotImplementedError`, as I don't see any point of raising exception in base class (which is a kind of interface or abstract probably) and overriding it in every subclass to return empty list. It's better to provide handy default in base class already. 
Also in future we may also want to refactor `OpenAPI` to implement `get_fuzzable_requests` and resign from openapi plugin, but it will break compatibility with current release by deleting one of plugins, so I opt for doing it later, ideally after moving to semantic versioning system.